### PR TITLE
ownCloud News support

### PR DESCRIPTION
**Pull request by dirb**
This PR implement the sync feature for the [ownCloud News App](https://apps.owncloud.com/content/show.php/News?content=168040), in response to this [request](https://github.com/akrennmair/newsbeuter/issues/134).

I have tested it a little, but a have a Slashdot feed of about two thousand items and I have not encountered a problem with that.

### CODE REVIEW

**Comment by Minoru** [comment]
No need to make separate commits to fix formatting—just amend the existing ones. (I mean `git rebase --interactive`.)

I'll try to review it properly in 6 hours or so. Thanks!

File: doc/newsbeuter.txt [hunk 1]
```diff
@@ -602,6 +602,7 @@ instead of +"+.)
 
 When you start newsbeuter, it will download the feeds that you configured
 in FeedHQ. Please take a closer look at the configuration commands for what 
+
```
Comment (Minoru) [inline doc/newsbeuter.txt]: A typo?

File: doc/configcommands.dsv [hunk 2]
```diff
@@ -99,8 +99,13 @@ ttrss-mode|[multi/single]|multi|Configures the mode in which Tiny Tiny RSS is us
 ttrss-password|<password>|""|Configures the password for use with Tiny Tiny RSS. Double quotes should be escaped, i.e. you should write +{backslash}"+ instead of +"+.|ttrss-password "here_goesAquote:{backslash}""
 ttrss-passwordfile|<path-to-file|""|A more secure alternative to the above, by storing your password elsewhere in your system.|ttrss-passwordfile "path-to-file"
 ttrss-url|<url>|""|Configures the URL where the Tiny Tiny RSS installation you want to use resides.|ttrss-url "http://example.com/ttrss/"
+ocnews-login|<username>|""|Sets the username for use with the ownCloud instance.|ocnews-login "user"
```
Comment (Minoru) [inline doc/configcommands.dsv]: It's "to use", not "for use". I see it's written that way in the rest of the file. You can amend your patch and leave the rest as it is, and I'll edit it later; or you can amend your patch and then make another commit to fix the rest yourself.

File: doc/newsbeuter.txt [hunk 3]
```diff
@@ -662,6 +663,34 @@ In order to use this mapping, you need to configure the flags that shall be used
 After that, use these flags when you edit flags for an article, and these articles
 will be starred resp. published.
 
+ownCloud News Synchronization
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For use the ownCloud News support you need to configure a few things
```
Comment (Minoru) [inline doc/newsbeuter.txt]: You forgot a full stop at the end of the sentence.

File: doc/newsbeuter.txt [hunk 4]
```diff
@@ -662,6 +663,34 @@ In order to use this mapping, you need to configure the flags that shall be used
...
+
+In addition, it requires username and password for authentication:
+
+	ocnews-login "user"
+	ocnews-password "password"
+
+ownCloud News API uses HTTP basic auth, therefore running ownCloud with SSL is highly
+recommended.
+
+In order to use this mapping, you need to configure the flags that shall be used:
```
Comment (Minoru) [inline doc/newsbeuter.txt]: It's not clear which mapping you mean here. Can you rephrase that? It's okay to repeat things you wrote in `configcommands.dsv`—we don't expect users to read the docs in their entirety before even starting the program.

File: src/ocnews_api.cpp [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+	                    + "/read?newestItemId=" + max;
+
+	return this->query(query, nullptr, "{}");
+}
+
+bool ocnews_api::mark_article_read(const std::string& guid, bool read) {
+	std::string query = "items/";
+	query += guid.substr(0, guid.find_first_of(":"));
+
+	if(read)
```
Comment (Minoru) [inline src/ocnews_api.cpp]: Whoops, no space after `if`?

**Comment by Minoru** [comment]
Looks okay (safe for my comments). You don't need to hurry with fixes as I don't know when I'll be able to actually test it, maybe this weekend—don't have time to set up an ownCloud instance right now.

Thanks!

File: doc/configcommands.dsv (hunk 2 above)
Comment (dirb) [inline doc/configcommands.dsv]: Sorry, I didn't know that. English isn't my first language, so I tend to just copy what I see other people do.

**Comment by dirb** [comment]
I will try to fix the things using `git rebase` rather than with individuals commits, I'm not good using git at this extent, so I hope no mess up things.

I also could let you to use my ownCloud instance, but it has a crappy internet conection so I don't know if it would be useful. Just email me if you're interested.

File: doc/configcommands.dsv (hunk 2 above)
Comment (Minoru) [inline doc/configcommands.dsv]: Ah. Well then, just fix what you can and I'll polish it later ;)

**Comment by Minoru** [comment]
If you end up messing up, just don't push—I can rebase things myself if need must (you'll still be listed as the author of the changes.)

Thanks for the offer, but I'd rather set up my own instance to make sure the code works not only with yours.

**Comment by dirb** [comment]
Done with the corrections.

**Comment by Minoru** [comment]
Just tested it. Everything seems fine save for one detail: if I star an article via OwnCloud's interface, it is not getting flagged in Newsbeuter. Can this be fixed?

**Comment by Minoru** [comment]
I also thought a bit more about when to release this and decided that it's okay to merge it for 2.10—this feature won't get much testing while it sits in HEAD anyway, so might as well unleash it upon the world right away :)

**Comment by dirb** [comment]
I have tried to do that!, I try to set the `starred` flag in the same way that the `read` status is set, but for some reason trying to call `rss_item::set_flags` inside of `rss_parser::fill_feed_items` doesn't work. Also, looking at the source code of other remote apis, it doesn't seen that they have that feature (at least on the json based ones), so I thought that it was fine without that.

I could put what I have try in another branch, so you could guide me about what happens, it seen to have something to do with the cache, and if it work it could be ported to the other remote apis.

**Comment by Minoru** [comment]
> Also, looking at the source code of other remote apis, it doesn't seen that they have that feature (at least on the json based ones)

Whoa, didn't know that! Let's fix it if we can, otherwise I'll open an issue for that and merge your code as it is.

> I could put what I have try in another branch, so you could guide me about what happens, it seen to have something to do with the cache, and if it work it could be ported to the other remote apis.

That would be great. I'm heading off to sleep, though, so I'll look at your code in about 10 hours.

**Comment by dirb** [comment]
This is what I refer to, [950e818](https://github.com/dirb/newsbeuter/commit/950e818a50fbebcd01209fe35ef37184e08da9c7#diff-9f23acca5027f2b0decf88c9e0fd8aaaR320).

**Comment by Minoru** [comment]
> it seen to have something to do with the cache

You are totally right! When you call `update_flags`, a query like `UPDATE rss_item SET flags = 's' WHERE guid = '10:1/bb042de0dd9d5febe786c32560afc358'` is run. But in this case, GUID is not set yet, so query doesn't update anything.

To solve this, simply move `set_guid` call (rss_parser.cpp:334) somewhere before the call to `update_flags`.

You can update other APIs if you wish. If not, I'll do it myself.

**Comment by dirb** [comment]
Yes, it works, but only if the item is already in the cache, it doesn't work when the item is fetched for the first time. In order to set the flags when the item is loaded for first time, I had to add the flags to the `INSERT INTO rss_item` query in [`src/cache.cpp`](https://github.com/dirb/newsbeuter/commit/723b469412185ae5374109959e802d8adaaa250d#diff-fe4e54b49aa936d63086740264cefc01R650).

I don't known much about the other apis, but I will try to add it to another api, like `oldreader` or `ttrss`.

**Comment by dirb** [comment]
I also would like to ask, should I merge this to the `ocnews-support` branch or wait it to be implemented in other apis?

**Comment by Minoru** [comment]
> Yes, it works, but only if the item is already in the cache

Oh, good catch!

Okay, here's what we're going to do. This PR already implements support for ownCloud News as good as for any other API, so I'm going to merge it, polish the docs and release it with 2.10.

As for starring the items, I've opened a new issue, #321, specifically about it. Once I'm done with 2.10 release (only 9 issues to go!), I'll figure out how existing code works and either merge your changes or explain why they're technically wrong.

> I also would like to ask, should I merge this to the ocnews-support branch or wait it to be implemented in other apis?

Start a new branch and work there. You can just wait for me to comment on #321, or you can just dive in and implement it for all other APIs yourself.

Thank you for your work!

**Comment by Minoru** [comment]
Merged via 8b1c2c4.
